streams/src/main/java/org/apache/kafka/streams/kstream/internals/suppress/KTableSuppressProcessor.java

```diff
@@ -17,70 +17,158 @@
+
+        final Bytes serializedKey = Bytes.wrap(keySerde.serializer().serialize(null, key));
+        final byte[] serializedValue = valueSerde.serializer().serialize(null, value);
+
+        buffer.put(time, serializedKey, new ContextualRecord<>(serializedValue, recordContext));
     }
 
     private long definedRecordTime(final K key) {
-        return suppress.getTimeDefinition().time(internalProcessorContext, key);
+        return timeDefinition.time(internalProcessorContext, key);
     }
 
-    @Override
-    public void close() {
+    private void enforceConstraints(final long streamTime) {
+        final long expiryTime = streamTime - suppressDurationMillis;
+        if (overCapacity() || buffer.minTimestamp() <= expiryTime) {
+
+            final Iterator<KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>>> iterator = buffer.iterator();
+
+            drainExpiredRecords(expiryTime, iterator);
+
+            if (overCapacity()) {
+                switch (bufferFullStrategy) {
+                    case EMIT:
+                        while (overCapacity() && (iterator.hasNext())) {
+                            evictAndEmit(iterator, iterator.next());
+                        }
+                        return;
+                    case SHUT_DOWN:
+                        throw new StreamsException(String.format(
+                            "%s buffer exceeded its max capacity [%d/%d] records and [%d/%d] bytes",
+                            internalProcessorContext.currentNode().name(),
+                            buffer.numRecords(), maxRecords,
+                            buffer.bufferSize(), maxBytes
+                        ));
+                }
+            }
+        }
     }
 
-    @Override
-    public String toString() {
-        return "KTableSuppressProcessor{" +
-            "suppress=" + suppress +
-            ", keySerde=" + keySerde +
-            ", valueSerde=" + valueSerde +
-            '}';
+    private void drainExpiredRecords(
+        final long expiryTime,
+        final Iterator<KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>>> iterator) {
+
+        while (iterator.hasNext() && buffer.minTimestamp() <= expiryTime) {
+            final KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>> next = iterator.next();
+            evictAndEmit(iterator, next);
+        }
+    }
+
+    private void evictAndEmit(final Iterator<KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>>> iterator,
+                              final KeyValue<TimeKey<Bytes>, ContextualRecord<byte[]>> next) {
+        final Bytes key = next.key.key();
+        final ContextualRecord<Change<V>> contextualRecord = new ContextualRecord<>(
+            valueSerde.deserializer().deserialize(null, next.value.value()),
+            next.value.recordContext()
+        );
+        // to be safe, forward before we delete
+        setNodeAndForward(key, contextualRecord);
```

Comment:
We've encountered some issues related to the ordering of this before: https://issues.apache.org/jira/browse/KAFKA-4492

Could you read that ticket and double check if flush-first-remove-later would not cause any issues for re-entrant puts on the same buffer (say, if we have a loop in the topology)?

Reply:
Interesting! That issue seems to be cache-specific: that two subsequent processors can be backed by the same cache (as in the join case).

I don't think loops are generally allowed in the subtopology, are they? If so, this code would indeed result in an infinite loop or possibly a concurrent modification exception.

I was concerned that the remove might be sent to the buffer's changelog record collector and maybe sent to the broker, and then some exception might happen before the forward, resulting in the record being forgotten upon restart.

I looked at some other processors, and they tend to do (logged) store operations first and then forward last. But then again, normal operations are forwarding a value that's a direct consequence of processing the _current_ record, which wouldn't have been committed and would therefore get re-processed upon restart.

But the buffer is forwarding some older record, which has already been committed. Reprocessing the new record (which caused the eviction the first time) won't cause us to remember the old record, which we were supposed to emit.

Under EOS, if we crash after the changelog update but before the forward, we'll be fine because the changelog update won't be visible (it'll be in an aborted transaction) on restart, so the buffer will go back to it's correct starting point for reprocessing the new record.

If we can't be sure that Streams subtopologies are acyclic, then I reckon we'd better swap these two lines and tell people they'd better use EOS if they want to be protected from all crash corruption (which I think is true anyway).
Otherwise, if subtopologies are acyclic, then I think it's better to leave it as is.

WDYT?

Reply:
I actually think that forward before delete is correct. Compare: https://issues.apache.org/jira/browse/KAFKA-5315 and the corresponding PR, that we never finished.